x/mongo/driver/auth/creds/aws_credential_providers/ec2_provider.go

```diff
@@ -0,0 +1,177 @@
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("response failure: %s", resp.Status)
+	}
+	return resp.Body, nil
+}
+
+func (e *Ec2Provider) getToken(ctx context.Context) (string, error) {
+	req, err := http.NewRequest(http.MethodPut, awsEC2URI+awsEC2TokenPath, nil)
+	if err != nil {
+		return "", err
+	}
+	const defaultEC2TTLSeconds = "30"
+	req.Header.Set("X-aws-ec2-metadata-token-ttl-seconds", defaultEC2TTLSeconds)
+
+	r, err := e.executeAWSHTTPRequest(ctx, req)
+	if err != nil {
+		return "", err
+	}
+	defer r.Close()
+
+	token, err := ioutil.ReadAll(r)
+	if err != nil {
+		return "", err
+	}
+	if len(token) == 0 {
+		return "", errors.New("unable to retrieve token from EC2 metadata")
+	}
+	return string(token), nil
+}
+
+func (e *Ec2Provider) getRoleName(ctx context.Context, token string) (string, error) {
+	req, err := http.NewRequest(http.MethodGet, awsEC2URI+awsEC2RolePath, nil)
+	if err != nil {
+		return "", err
+	}
+	req.Header.Set("X-aws-ec2-metadata-token", token)
+
+	r, err := e.executeAWSHTTPRequest(ctx, req)
+	if err != nil {
+		return "", err
+	}
+	defer r.Close()
+
+	role, err := ioutil.ReadAll(r)
+	if err != nil {
+		return "", err
+	}
+	if len(role) == 0 {
+		return "", errors.New("unable to retrieve role_name from EC2 metadata")
+	}
+	return string(role), nil
+}
+
+func (e *Ec2Provider) getCredentials(ctx context.Context, token string, role string) (credentials.Value, time.Time, error) {
+	v := credentials.Value{ProviderName: Ec2ProviderName}
+
+	pathWithRole := awsEC2URI + awsEC2RolePath + role
+	req, err := http.NewRequest(http.MethodGet, pathWithRole, nil)
+	if err != nil {
+		return v, time.Time{}, err
+	}
+	req.Header.Set("X-aws-ec2-metadata-token", token)
+	resp, err := e.executeAWSHTTPRequest(ctx, req)
+	if err != nil {
+		return v, time.Time{}, err
+	}
+	defer resp.Close()
+
+	var ec2Resp struct {
+		AccessKeyID     string    `json:"AccessKeyId"`
+		SecretAccessKey string    `json:"SecretAccessKey"`
+		Token           string    `json:"Token"`
+		Expiration      time.Time `json:"Expiration"`
+	}
+
+	err = json.NewDecoder(resp).Decode(&ec2Resp)
+	if err != nil {
+		return v, time.Time{}, err
+	}
+
+	v.AccessKeyID = ec2Resp.AccessKeyID
+	v.SecretAccessKey = ec2Resp.SecretAccessKey
+	v.SessionToken = ec2Resp.Token
+
+	return v, ec2Resp.Expiration, nil
+}
+
+// RetrieveWithContext retrieves the keys from the AWS service.
+func (e *Ec2Provider) RetrieveWithContext(ctx context.Context) (credentials.Value, error) {
+	v := credentials.Value{ProviderName: Ec2ProviderName}
+
+	token, err := e.getToken(ctx)
+	if err != nil {
+		return v, err
+	}
+
+	role, err := e.getRoleName(ctx, token)
+	if err != nil {
+		return v, err
+	}
+
+	v, exp, err := e.getCredentials(ctx, token, role)
+	if err != nil {
+		return v, err
+	}
+	if !v.HasKeys() {
+		return v, errors.New("failed to retrieve EC2 keys")
+	}
+	e.expiration = exp.Add(-e.expiryWindow)
+
+	return v, err
```

Comment:
`err` here can only be `nil`, so it's clearer to just return `nil` to avoid confusion about what the expected return values are.

```suggestion
	return v, nil
```